mysql-test/suite/innodb/t/undo_leak.test

```diff
@@ -24,3 +24,58 @@ SELECT NAME, FILE_SIZE FROM information_schema.innodb_sys_tablespaces WHERE SPAC
 let $restart_parameters=;
 --source include/restart_mysqld.inc
 SELECT NAME, FILE_SIZE FROM information_schema.innodb_sys_tablespaces WHERE SPACE = 0;
+
+--echo #
+--echo # MDEV-35398 Improve shrinking of system tablespace
+--echo #
+# Make pre-10.6.13 version data directory which skips the
+# purging of cached undo log records
+let $restart_parameters=--debug_dbug=d,skip_cached_undo;
+--source include/restart_mysqld.inc
+set global innodb_file_per_table= 0;
+create table t1(f1 int not null)engine=innodb;
+create table t2(f2 int not null)engine=innodb;
+INSERT INTO t1 SELECT seq FROM seq_1_to_32768;
+INSERT INTO t2 SELECT seq FROM seq_1_to_32768;
+
+--echo # Insert 63 transaction which has undo cached records
+--disable_query_log
+let $c = 63;
+while ($c)
+{
+  connect (con$c,localhost,root,,);
+  dec $c;
+}
+
+let $c = 63;
+while ($c)
+{
+  connection con$c;
+  START TRANSACTION;
+  eval INSERT INTO t1 select seq from seq_1_to_4096;
+  dec $c;
+}
```

Comment:
These two loops can be combined. `connect` implies a change of the current `connection`.